classes/task/metadata_refresh.php

```diff
@@ -0,0 +1,147 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Auth SAML2 metadata refresh scheduled task.
+ *
+ * @package    auth_saml2
+ * @author     Sam Chaffee
+ * @copyright  Copyright (c) 2017 Blackboard Inc. (http://www.blackboard.com)
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+namespace auth_saml2\task;
+
+use auth_saml2\config;
+use auth_saml2\metadata_fetcher;
+use auth_saml2\metadata_parser;
+use auth_saml2\metadata_writer;
+
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Auth SAML2 metadata refresh scheduled task.
+ *
+ * @package    auth_saml2
+ * @copyright  Copyright (c) 2017 Blackboard Inc. (http://www.blackboard.com)
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class metadata_refresh extends \core\task\scheduled_task {
+
+    /**
+     * @var config
+     */
+    private $config;
+
+    /**
+     * @var metadata_fetcher
+     */
+    private $fetcher;
+
+    /**
+     * @var metadata_parser
+     */
+    private $parser;
+
+    /**
+     * @var metadata_writer
+     */
+    private $writer;
+
+    public function get_name() {
+        return get_string('taskmetadatarefresh', 'auth_saml2');
+    }
+
+    public function execute() {
+        // Check config to see that the metadata is a URL, not XML. Return if not a URL.
+        if (!$this->config instanceof config) {
+            $this->config = new config();
+        }
+        if (substr($this->config->idpmetadata, 0, 8) != 'https://'
+                && substr($this->config->idpmetadata, 0, 7) != 'http://') {
+            // Not a link so nothing to refresh.
+            mtrace('IDP metadata config not a URL, nothing to refresh.');
+            return;
+        }
+        // Fetch the metadata.
+        if (!$this->fetcher instanceof metadata_fetcher) {
+            $this->fetcher = new metadata_fetcher();
+        }
+
+        try {
+            $rawxml = $this->fetcher->fetch($this->config->idpmetadata);
```

Comment:
@brendanheywood, I'm re-thinking these bits. At first, I thought that we wouldn't want the task to fail and reschedule itself, but it may be better to let it go ahead and do that. In that case I wouldn't bother wrapping those calls in try/catch blocks. Instead, just let the exceptions bubble up and have the task scheduler handle.